hail/src/main/scala/is/hail/expr/ir/agg/StagedBlockLinkedList.scala

```diff
@@ -0,0 +1,242 @@
+      Region.storeInt(nodeType.fieldOffset(n, 1), count),
+      nodeType.setFieldMissing(n, 2))
+
+  private def pushPresent(n: Node, store: Code[Long] => Code[Unit]): Code[Unit] =
+    Code(
+      if(elemType.required) Code._empty else bufferType.setElementPresent(buffer(n), count(n)),
+      store(bufferType.elementOffset(buffer(n), capacity(n), count(n))),
+      incrCount(n))
+
+  private def pushMissing(n: Node): Code[Unit] = {
+    if(elemType.required)
+      Code._fatal("cannot push missing onto list with required elements")
+    else
+      Code(
+        bufferType.setElementMissing(buffer(n), count(n)),
+        incrCount(n))
+  }
+
+  private def allocateNode(r: Code[Region], cap: Code[Int]): (Code[Unit], Node) = {
+    val setup = Code(
+      tmpNode := r.allocate(nodeType.alignment, nodeType.byteSize),
+      initNode(tmpNode,
+        buf = r.allocate(bufferType.contentsAlignment, bufferType.contentsByteSize(cap)),
+        count = 0),
+      bufferType.stagedInitialize(buffer(tmpNode), cap))
+    (setup, tmpNode)
+  }
+
+  def initWithCapacity(r: Code[Region], initialCap: Code[Int]): Code[Unit] = {
+    val (setup, n) = allocateNode(r, initialCap)
+    Code(setup,
+      firstNode := n,
+      lastNode := n,
+      totalCount := 0)
+  }
+
+  def init(r: Code[Region]): Code[Unit] =
+    initWithCapacity(r, defaultBlockCap)
+
+  private[agg] def pushNewBlockNode(r: Code[Region], cap: Code[Int]): Code[Unit] = {
+    val (setup, newNode) = allocateNode(r, cap)
+    Code(setup,
+      setNext(lastNode, newNode),
+      lastNode := newNode)
+  }
+
+  private[agg] def foreachNode(f: Node => Code[Unit]): Code[Unit] = {
+    Code(
+      tmpNode := firstNode,
+      p := true,
+      Code.whileLoop(p,
+        f(tmpNode),
+        p := hasNext(tmpNode),
+        tmpNode := next(tmpNode)))
+  }
+
+  def foreach(f: EmitTriplet => Code[Unit]): Code[Unit] = {
+    def et(n: Node, i: Code[Int]) =
+      EmitTriplet(Code._empty,
+        bufferType.isElementMissing(buffer(n), i),
+        Region.loadIRIntermediate(elemType)(
+          bufferType.elementOffset(buffer(n), capacity(n), i)))
+    foreachNode { n => Code(
+      i := 0,
+      Code.whileLoop(i < count(n),
+        f(et(n, i)),
+        i := i + 1))
+    }
+  }
+
+  def push(r: Code[Region], elt: EmitTriplet): Code[Unit] = {
+    Code(
+      (count(lastNode) >= capacity(lastNode)).orEmpty(
+        pushNewBlockNode(r, defaultBlockCap)), // push a new block if lastNode is full
+      elt.setup,
+      elt.m.mux(
+        pushMissing(lastNode),
+        pushPresent(lastNode, { dst =>
+          if(elemType.isPrimitive)
+            Region.storePrimitive(elemType, dst)(elt.value)
+          else
+            StagedRegionValueBuilder.deepCopy(fb, r, elemType, elt.value, dst)
+        })),
+      totalCount := totalCount + 1)
+  }
+
+  def append(r: Code[Region], bll: StagedBlockLinkedList): Code[Unit] = {
+    assert(bll.elemType.isOfType(elemType))
+    bll.foreach(push(r, _))
```

Comment:
Similarly to `initWithDeepCopy`, could this copy each buffer of `bll` as a unit? Also, might there be cases where we could just steal the nodes of `bll`, and avoid copies entirely?

Reply:
i'm really not sure how the region lifetimes interact when there are two aggregators, so i'm hesitant to try to share/steal pointers. but it would definitely be a desirable way to do append